### PR TITLE
[Perf][Kernel][Quantization] Fuse ReLU2 with static FP8 activation quantization

### PR DESCRIPTION
**Pull request by samnordmann**
## Purpose

Fuse BF16 ReLU2 with static per-tensor FP8 activation quantization for
compatible Nemotron-H MLP down-projections. The Triton producer preserves the
observable BF16 rounding boundary and returns a `QuantizedActivation` through
the generic `maybe_fused_act_quant` interface, allowing the consumer to use its
checkpoint input scale without quantizing twice.

This implements the activation-quantization work tracked by #43501, integrates
with the manual-fusion interface from #51415, and uses the consumer contract
from #44260. It supersedes closed draft #53791, which was an environment-gated,
model-local prototype. No other open PR found in the duplicate-work search
implements this ReLU2-to-static-FP8 producer/consumer path.

The final dispatch is capability-driven:

- Nemotron-H MLPs call `maybe_fused_act_quant` unconditionally; unsupported
  inputs fall back to the existing ReLU2 path.
- The producer is registered only on CUDA SM90+ for consumers advertising
  `kFp8StaticTensorSym`.
- Its support predicate requires a contiguous BF16 CUDA input and a scalar
  FP32 `input_scale` on the same device.
- LoRA linear wrappers declare that their adapter branch requires the original
  activation, so the generic consumer contract suppresses pre-quantized input
  for them.
- Selection is independent of compilation mode, tensor-parallel degree, and
  the standalone `relu2` / `quant_fp8` custom-op choices.

Maintainer follow-up `c1fc034a90` centralized producer eligibility and LoRA
handling in the generic contract. The branch merges upstream `main` at
`c268198715`; the effective PR diff is 12 files (`+459/-32`).

## Test Plan

```text
pre-commit run --files $(git diff --name-only upstream/main...HEAD)

python -m pytest -q \
  tests/kernels/test_relu2_fp8_quant.py \
  tests/kernels/test_fused_quant_activation.py \
  tests/compile/passes/test_silu_mul_quant_manual_fusion.py \
  tests/model_executor/test_nemotron_h_quantization.py \
  tests/fusion/test_quant_activation_contract.py \
  tests/lora/test_layers.py::test_lora_linear_requires_unquantized_input \
  tests/quantization/test_modelopt.py::test_modelopt_fp8_updates_weight_dims_after_transpose \
  tests/quantization/test_modelopt.py::test_modelopt_linear_algos_table_matches_resolve \
  tests/quantization/test_modelopt.py::test_modelopt_mixed_precision_dispatches_every_linear_algo \
  tests/quantization/test_modelopt.py::test_modelopt_mixed_precision_dispatches_linear_layer
```

GPU validation used one NVIDIA GB200 node and
`vllm/vllm-openai:cu129-nightly-73029d42441321b631779db3475031f5ec26dd6c`
(PyTorch `2.13.0+cu129`). The reviewed Python/Triton source was overlaid on that
immutable runtime; this PR does not change C++/CUDA extension source.

## Test Result

- At merge head `dc81524d9f`, the complete local pre-commit file set passes,
  including ruff, formatting, typos, Python 3.10 mypy, SPDX, lazy-import,
  forbidden-import, CUDA-API, and configuration checks. `git diff --check`
  also passes. The fresh [GitHub pre-commit
  run](https://github.com/vllm-project/vllm/actions/runs/34295261882) passes.
- Full [PR CI build 87766](https://buildkite.com/vllm/ci/builds/87766) at
  `6818a6d057` included the maintainer refactor and passed all PR-relevant
  fusion, kernel, quantization, model-executor, and LoRA jobs. Its remaining
  multimodal failures reproduced on the [exact-base main build
  87764](https://buildkite.com/vllm/ci/builds/87764):
  `CohereLabs/North-Micro-Vision-Instruct` exposed a nested
  `rope_parameters=None`. Upstream #55949 fixed that issue, and the affected
  jobs pass in [main build 87769](https://buildkite.com/vllm/ci/builds/87769).
  The current merge includes that fix; a fresh full Buildkite run has not yet
  been triggered.
- Lyris job `2891596` (NVIDIA GB200, exit 0) validated exact source
  `ae2a8140de`: `225 passed`. All 65,536 BF16 encodings, including 254 NaN
  encodings and both infinities, produced zero raw-byte mismatches at all 48
  checkpoint scales for the direct producer, `torch.compile(fullgraph=True)`,
  and CUDA-graph replay. Producer metadata also matched the downstream
  `QuantizedActivation` contract for every scale. The dynamic-token fullgraph
  test reused one compiled frame across 17, 31, and 63 tokens. The Triton
  producer implementation is unchanged at the current head; subsequent changes
  affect dispatch and consumer plumbing.
- Lyris job `2891597` (NVIDIA GB200, exit 0) loaded the 550B Nemotron-H
  checkpoint with O2/Inductor at TP1/DP4/EP4. At `M=1` and `M=128`, the loaded
  shared-expert producer, public ModelOpt `QuantizedActivation` consumer, and
  FlashInfer down-projection controls matched raw bytes across repeat,
  pointer-order, scale-identity, and native/fused-order checks. Four fixed
  serving prompts each completed one token. This is a loaded
  producer/consumer and basic serving smoke, not an output-quality, sustained
  stability, or end-to-end performance claim.
- Lyris job `2856442`: full 550B load, Inductor compilation, piecewise/full
  CUDA-graph capture, and serving smoke. The producer was selected at all 48
  shared-expert sites per rank and four requests completed without a graph
  break or traceback. This used predecessor source and is retained only as an
  integration smoke, not an accuracy result.

The exhaustive bit-exact claim is specifically against the default,
separately compiled O2 native ReLU2 and QuantFP8 boundary and the loaded
FlashInfer consumer used above. It is not an all-custom-op-dispatch or
end-to-end model-output claim.

## Performance

Lyris job `2891596`, hidden size 10,240, one recorded checkpoint scale:

| Tokens | Deployed separate O2 ops | Joint Inductor diagnostic | Manual producer | Manual vs. deployed | Manual vs. joint |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 16,384 | 475.686 us | 119.156 us | 78.083 us | 6.09x | 1.53x |
| 32,768 | 945.178 us | 234.952 us | 149.843 us | 6.31x | 1.57x |

The manual producer was byte-exact to the deployed O2 boundary. The joint
Inductor diagnostic was intentionally not used as a parity oracle: removing
the observable BF16 intermediate rounding changed 1,116,386 and 2,232,217
output bytes at the two shapes, respectively.

The isolated kernel-boundary comparison uses the deployed O2 baseline (native
ReLU2 and QuantFP8 operators compiled separately) and a fullgraph-compiled
manual producer. Compilation and autotuning finish before capture; seven
counterordered samples of 25 CUDA-graph replays are collected per shape and the
median is reported. This is not an end-to-end model latency measurement.

The prior full-model campaign was measured on predecessor source
[`6a9bb4080b`](https://github.com/samnordmann/vllm/commit/6a9bb4080b1757fcbd6cb909edbd1429eb74aa89),
so it remains directional rather than a current-head claim:

- shared-MLP kernel time at `M=32768`: `3655.27 -> 2968.37 us` (`-18.79%`);
- prefill-heavy ISL/OSL `50000/1`, six counterordered pairs: request
  throughput `1.4311 -> 1.4505 req/s` (`+1.34%`), paired 95% interval
  `+0.05%` to `+2.64%`;
- long-output `50000/2048`, six pairs: mean TPOT central change `-0.01%`, with
  paired 95% interval `-5.68%` to `+5.99%`. This is underpowered and does not
  establish a decode improvement or rule out a regression.

## AI assistance

OpenAI Codex was used to review, test, and edit this PR. Claude Code's supplied
independent review was also critically audited. The author remains responsible
for reviewing, understanding, and defending the change.

### CODE REVIEW

File: vllm/model_executor/models/nemotron_h.py [hunk 1]
```diff
@@ -115,10 +130,35 @@ def __init__(
...
+            if enable_relu2_fp8_quant
+            and current_platform.is_cuda()
+            and not self.act_fn.enabled()
+            and not QuantFP8.enabled()
+            and Bf16ReLUSquaredStaticFp8Quant.is_supported_in_current_config()
+            and get_tensor_model_parallel_world_size() == 1
+            and getattr(self.down_proj, "input_quant_key", None) == kFp8StaticTensorSym
+            and hasattr(self.down_proj, "input_scale")
+            else None
+        )
```
Comment (mgoin) [inline vllm/model_executor/models/nemotron_h.py]: Thanks for using the input_quant_key and QuantizedActivation structure. We are just about to land the manual version of `ActivationQuantFusionPass` using this here https://github.com/vllm-project/vllm/pull/51415, so if you wouldn't mind integrating with that I think it'd be best

File: vllm/model_executor/layers/fusion/relu2_fp8_quant.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        assert x.dtype == torch.bfloat16
+        assert x.is_contiguous()
+        assert scale.dtype == torch.float32
+        assert x.device == scale.device
+        assert scale.numel() == 1
+        output = torch.empty_like(x, dtype=current_platform.fp8_dtype())
+        if x.numel() == 0:
+            return output
+        block_size = 2048
+        _bf16_relu2_static_fp8_quant_kernel[(triton.cdiv(x.numel(), block_size),)](
```
Comment (mgoin) [inline vllm/model_executor/layers/fusion/relu2_fp8_quant.py]: Probably worth tuning/capping this at least a little bit for different sizes

File: tests/kernels/test_relu2_fp8_quant.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import pytest
+import torch
+
+from vllm.config import CompilationConfig, VllmConfig, set_current_vllm_config
+from vllm.config.vllm import OptimizationLevel
+from vllm.model_executor.layers.activation import ReLUSquaredActivation
+from vllm.model_executor.layers.fusion.relu2_fp8_quant import (
+    Bf16ReLUSquaredStaticFp8Quant,
```
Comment (mgoin) [inline tests/kernels/test_relu2_fp8_quant.py]: nit: would just remove the Bf16 for the name overall

File: tests/kernels/test_relu2_fp8_quant.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+)
+from vllm.model_executor.layers.quantization.input_quant_fp8 import QuantFP8
+from vllm.model_executor.layers.quantization.utils.quant_utils import GroupShape
+from vllm.platforms import current_platform
+
+requires_cuda = pytest.mark.skipif(
+    not current_platform.is_cuda(), reason="This test requires CUDA"
+)
```
Comment (mgoin) [inline tests/kernels/test_relu2_fp8_quant.py]: Also requires >= sm90 `current_platform.current_platform.has_device_capability(90)`

File: vllm/model_executor/models/nemotron_h.py (hunk 1 above)
Comment (samnordmann) [reply vllm/model_executor/models/nemotron_h.py]: Ok, I merged #51415 and moved the producr to `maybe_fused_act_quant` registry keyed by the downstream `input_quant_key`

File: vllm/model_executor/layers/fusion/relu2_fp8_quant.py (hunk 2 above)
Comment (samnordmann) [reply vllm/model_executor/layers/fusion/relu2_fp8_quant.py]: Done. `BLOCK_SIZE` now uses `min(next_power_of_2(x.shape[-1]), 2048)`, and `num_warps` scales with it up to 4

File: vllm/model_executor/models/nemotron_h.py [hunk 5]
```diff
@@ -115,10 +130,26 @@ def __init__(
...
+            enable_relu2_fp8_quant
+            and current_platform.is_cuda()
+            and not self.act_fn.enabled()
+            and not QuantFP8.enabled()
+            and current_platform.has_device_capability(90)
+            and is_relu_squared_static_fp8_quant_config_supported()
+            and get_tensor_model_parallel_world_size() == 1
+            and getattr(self.down_proj, "input_quant_key", None) == kFp8StaticTensorSym
+            and hasattr(self.down_proj, "input_scale")
+        )
```
Comment (mgoin) [inline vllm/model_executor/models/nemotron_h.py]: I think we could put most this logic into _relu_squared_static_fp8_quant_supported, any reason not to? Do we really need to care about the custom op choices?

**Review by mgoin — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
